# Working on Real Project with Python 

# (A part of Big Data Analysis)

-----

# The Weather Dataset
***
Here, 
The Weather Dataset is a time-series data set with per-hour information about the weather conditions at a particular location. It records Temperature, Dew Point Temperature, Relative Humidity, Wind Speed, Visibility, Pressure, and Conditions.


This data is available as a CSV file. We are going to analyze this data set using the Pandas DataFrame.

In [140]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [141]:
df = pd.read_csv('Weather Data.csv')
df.head()

,Date/Time,Temp_C,Dew Point Temp_C,Rel Hum_%,Wind Speed_km/h,Visibility_km,Press_kPa,Weather
0,1/1/2012 0:00,-1.8,-3.9,86,4,8.0,101.24,Fog
1,1/1/2012 1:00,-1.8,-3.7,87,4,8.0,101.24,Fog
2,1/1/2012 2:00,-1.8,-3.4,89,7,4.0,101.26,"Freezing Drizzle,Fog"
3,1/1/2012 3:00,-1.5,-3.2,88,6,4.0,101.27,"Freezing Drizzle,Fog"
4,1/1/2012 4:00,-1.5,-3.3,88,7,4.8,101.23,Fog


# How to Analyze DataFrames ?

# .head()
It shows the first N rows in the data (by default, N=5).

In [142]:
df.head()

,Date/Time,Temp_C,Dew Point Temp_C,Rel Hum_%,Wind Speed_km/h,Visibility_km,Press_kPa,Weather
0,1/1/2012 0:00,-1.8,-3.9,86,4,8.0,101.24,Fog
1,1/1/2012 1:00,-1.8,-3.7,87,4,8.0,101.24,Fog
2,1/1/2012 2:00,-1.8,-3.4,89,7,4.0,101.26,"Freezing Drizzle,Fog"
3,1/1/2012 3:00,-1.5,-3.2,88,6,4.0,101.27,"Freezing Drizzle,Fog"
4,1/1/2012 4:00,-1.5,-3.3,88,7,4.8,101.23,Fog


In [143]:
df.columns = df.columns.str.lower().str.replace(' ','_')

# .shape

It shows the total no. of rows and no. of columns of the dataframe

In [144]:
df.shape

(8784, 8)

# .index
This attribute provides the index of the dataframe

In [145]:
df.index

RangeIndex(start=0, stop=8784, step=1)

# .columns
It shows the name of each column

In [146]:
df.columns

Index(['date/time', 'temp_c', 'dew_point_temp_c', 'rel_hum_%',
       'wind_speed_km/h', 'visibility_km', 'press_kpa', 'weather'],
      dtype='object')

# .dtypes
It shows the data-type of each column 

In [147]:
df.dtypes

date/time            object
temp_c              float64
dew_point_temp_c    float64
rel_hum_%             int64
wind_speed_km/h       int64
visibility_km       float64
press_kpa           float64
weather              object
dtype: object

# .unique()
In a column, it shows all the unique values. 
It can be applied on a single column only, not on the whole dataframe.

In [148]:
df.weather.unique()

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

# .nunique()
It shows the total no. of unique values in each column. 
It can be applied on a single column as well as on whole dataframe.

In [149]:
df.weather.nunique()

50

# .count 
It shows the total no. of non-null values in each column. 
It can be applied on a single column as well as on whole dataframe.

In [150]:
df.count()

date/time           8784
temp_c              8784
dew_point_temp_c    8784
rel_hum_%           8784
wind_speed_km/h     8784
visibility_km       8784
press_kpa           8784
weather             8784
dtype: int64

# .value_counts
In a column, it shows all the unique values with their count. It can be applied on single column only.

In [151]:
df.weather.value_counts()

weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

# .info()
Provides basic information about the dataframe.

In [152]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   date/time         8784 non-null   object 
 1   temp_c            8784 non-null   float64
 2   dew_point_temp_c  8784 non-null   float64
 3   rel_hum_%         8784 non-null   int64  
 4   wind_speed_km/h   8784 non-null   int64  
 5   visibility_km     8784 non-null   float64
 6   press_kpa         8784 non-null   float64
 7   weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


-----

# Q) 1.  Find all the unique 'Wind Speed' values in the data.

In [153]:
df

,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa,weather
0,1/1/2012 0:00,-1.8,-3.9,86,4,8.0,101.24,Fog
1,1/1/2012 1:00,-1.8,-3.7,87,4,8.0,101.24,Fog
2,1/1/2012 2:00,-1.8,-3.4,89,7,4.0,101.26,"Freezing Drizzle,Fog"
3,1/1/2012 3:00,-1.5,-3.2,88,6,4.0,101.27,"Freezing Drizzle,Fog"
4,1/1/2012 4:00,-1.5,-3.3,88,7,4.8,101.23,Fog
...,...,...,...,...,...,...,...,...
8779,12/31/2012 19:00,0.1,-2.7,81,30,9.7,100.13,Snow
8780,12/31/2012 20:00,0.2,-2.4,83,24,9.7,100.03,Snow
8781,12/31/2012 21:00,-0.5,-1.5,93,28,4.8,99.95,Snow
8782,12/31/2012 22:00,-0.2,-1.8,89,28,9.7,99.91,Snow


In [154]:
df['wind_speed_km/h'].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2])

# Q) 2. Find the number of times when the 'Weather is exactly Clear'.

In [155]:
# value_counts()
df.weather.value_counts()

weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

In [156]:
# Filtering
#data.head(2)
df[df['weather'] == 'Clear']['weather'].value_counts()

weather
Clear    1326
Name: count, dtype: int64

In [157]:
# groupby()
#data.head(2)
df.groupby('weather').get_group('Clear')


,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa,weather
67,1/3/2012 19:00,-16.9,-24.8,50,24,25.0,101.74,Clear
114,1/5/2012 18:00,-7.1,-14.4,56,11,25.0,100.71,Clear
115,1/5/2012 19:00,-9.2,-15.4,61,7,25.0,100.80,Clear
116,1/5/2012 20:00,-9.8,-15.7,62,9,25.0,100.83,Clear
117,1/5/2012 21:00,-9.0,-14.8,63,13,25.0,100.83,Clear
...,...,...,...,...,...,...,...,...
8646,12/26/2012 6:00,-13.4,-14.8,89,4,25.0,102.47,Clear
8698,12/28/2012 10:00,-6.1,-8.6,82,19,24.1,101.27,Clear
8713,12/29/2012 1:00,-11.9,-13.6,87,11,25.0,101.31,Clear
8714,12/29/2012 2:00,-11.8,-13.1,90,13,25.0,101.33,Clear


The number of times the clear weather weather appears is 1326 times

# Q) 3. Find the number of times when the 'Wind Speed was exactly 4 km/h'.

In [158]:
df[df['wind_speed_km/h'] == 4]['wind_speed_km/h'].value_counts()



wind_speed_km/h
4    474
Name: count, dtype: int64

the number of times when the 'Wind Speed was exactly 4 km/h' is 474 times.

# Q. 4) Find out all the Null Values in the data.

In [159]:
df.isna().sum()

date/time           0
temp_c              0
dew_point_temp_c    0
rel_hum_%           0
wind_speed_km/h     0
visibility_km       0
press_kpa           0
weather             0
dtype: int64

there is no null values in the data

# Q. 5) Rename the column name 'Weather' of the dataframe to 'Weather Condition'.

In [160]:
df.head(2)

,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa,weather
0,1/1/2012 0:00,-1.8,-3.9,86,4,8.0,101.24,Fog
1,1/1/2012 1:00,-1.8,-3.7,87,4,8.0,101.24,Fog


In [161]:
df.rename(columns = {'weather':'weather_conditions'},inplace = True)

In [162]:
df.head(1)

,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa,weather_conditions
0,1/1/2012 0:00,-1.8,-3.9,86,4,8.0,101.24,Fog


# Q.6) What is the mean 'Visibility' ?

In [163]:
df.visibility_km.mean().round(2)

np.float64(27.66)

The mean of visibility is 27.66

# Q. 7) What is the Standard Deviation of 'Pressure'  in this data?

In [164]:
df.press_kpa.std()

0.8440047459486474

the Standard Deviation of 'Pressure'  in this data is 0.8

# Q. 8) Whats is the Variance of 'Relative Humidity' in this data ?

In [165]:
df['rel_hum_%'].var()

286.2485501984998

the Variance of 'Relative Humidity' in this data is 286

# Q. 9) Find all instances when 'Snow' was recorded.

In [166]:
# value_counts()
#data.head(2)
df.head(1)
df.weather_conditions.value_counts()

weather_conditions
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle     

In [167]:
#Filtering
df[df['weather_conditions'] == 'Snow'].weather_conditions.value_counts()


weather_conditions
Snow    390
Name: count, dtype: int64

In [168]:
# str.contains
df[df.weather_conditions.str.contains('Snow')]

,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa,weather_conditions
41,1/2/2012 17:00,-2.1,-9.5,57,22,25.0,99.66,Snow Showers
44,1/2/2012 20:00,-5.6,-13.4,54,24,25.0,100.07,Snow Showers
45,1/2/2012 21:00,-5.8,-12.8,58,26,25.0,100.15,Snow Showers
47,1/2/2012 23:00,-7.4,-14.1,59,17,19.3,100.27,Snow Showers
48,1/3/2012 0:00,-9.0,-16.0,57,28,25.0,100.35,Snow Showers
...,...,...,...,...,...,...,...,...
8779,12/31/2012 19:00,0.1,-2.7,81,30,9.7,100.13,Snow
8780,12/31/2012 20:00,0.2,-2.4,83,24,9.7,100.03,Snow
8781,12/31/2012 21:00,-0.5,-1.5,93,28,4.8,99.95,Snow
8782,12/31/2012 22:00,-0.2,-1.8,89,28,9.7,99.91,Snow


all the instances when 'Snow' was recorded is 583

# Q. 10) Find all instances when 'Wind Speed is above 24' and 'Visibility is 25'.

In [169]:
df.head(2)

,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa,weather_conditions
0,1/1/2012 0:00,-1.8,-3.9,86,4,8.0,101.24,Fog
1,1/1/2012 1:00,-1.8,-3.7,87,4,8.0,101.24,Fog


In [170]:
df[(df['wind_speed_km/h'] > 24) & (df['visibility_km'] == 25)]['weather_conditions'].value_counts()



weather_conditions
Cloudy              99
Mostly Cloudy       94
Mainly Clear        52
Rain                16
Snow                15
Clear               12
Rain Showers         9
Snow Showers         8
Rain,Snow            2
Rain,Snow Grains     1
Name: count, dtype: int64

# Q. 11) What is the Mean value of each column against each 'Weather Conditon' ?

In [171]:
df.groupby('weather_conditions')['visibility_km'].mean().round(2)

weather_conditions
Clear                                      30.15
Cloudy                                     26.63
Drizzle                                    17.93
Drizzle,Fog                                 5.26
Drizzle,Ice Pellets,Fog                     4.00
Drizzle,Snow                               10.50
Drizzle,Snow,Fog                            5.51
Fog                                         6.25
Freezing Drizzle                            9.20
Freezing Drizzle,Fog                        5.27
Freezing Drizzle,Haze                       2.67
Freezing Drizzle,Snow                       5.87
Freezing Fog                                0.65
Freezing Rain                               8.24
Freezing Rain,Fog                           7.55
Freezing Rain,Haze                          2.40
Freezing Rain,Ice Pellets,Fog               8.00
Freezing Rain,Snow Grains                   4.80
Haze                                        7.83
Mainly Clear                               34.26
M

# Q. 12) What is the Minimum & Maximum value of each column against each 'Weather Conditon' ?

In [173]:
df.groupby('weather_conditions').min()

,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa
weather_conditions,,,,,,,
Clear,1/11/2012 1:00,-23.3,-28.5,20,0,11.3,99.52
Cloudy,1/1/2012 17:00,-21.4,-26.8,18,0,11.3,98.39
Drizzle,1/23/2012 21:00,1.1,-0.2,74,0,6.4,97.84
"Drizzle,Fog",1/23/2012 20:00,0.0,-1.6,85,0,1.0,98.65
"Drizzle,Ice Pellets,Fog",12/17/2012 9:00,0.4,-0.7,92,20,4.0,100.79
"Drizzle,Snow",12/17/2012 15:00,0.9,0.1,92,9,9.7,100.63
"Drizzle,Snow,Fog",12/18/2012 21:00,0.3,-0.1,92,7,2.4,97.79
Fog,1/1/2012 0:00,-16.0,-17.2,80,0,0.2,98.31
Freezing Drizzle,1/13/2012 10:00,-9.0,-12.2,78,6,4.8,98.44


In [176]:
df.groupby('weather_conditions').max()

,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa
weather_conditions,,,,,,,
Clear,9/9/2012 5:00,32.8,20.4,99,33,48.3,103.63
Cloudy,9/9/2012 23:00,30.5,22.6,99,54,48.3,103.65
Drizzle,9/30/2012 3:00,18.8,17.7,96,30,25.0,101.56
"Drizzle,Fog",9/30/2012 2:00,19.9,19.1,100,28,9.7,102.07
"Drizzle,Ice Pellets,Fog",12/17/2012 9:00,0.4,-0.7,92,20,4.0,100.79
"Drizzle,Snow",12/19/2012 18:00,1.2,0.2,95,19,11.3,101.15
"Drizzle,Snow,Fog",12/22/2012 3:00,1.1,0.6,98,32,9.7,100.15
Fog,9/22/2012 0:00,20.8,19.6,100,22,9.7,103.04
Freezing Drizzle,2/1/2012 5:00,-2.3,-3.3,93,26,12.9,101.02


# Q. 13) Show all the Records where Weather Condition is Fog.

In [178]:
df[df.weather_conditions == 'Fog']

,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa,weather_conditions
0,1/1/2012 0:00,-1.8,-3.9,86,4,8.0,101.24,Fog
1,1/1/2012 1:00,-1.8,-3.7,87,4,8.0,101.24,Fog
4,1/1/2012 4:00,-1.5,-3.3,88,7,4.8,101.23,Fog
5,1/1/2012 5:00,-1.4,-3.3,87,9,6.4,101.27,Fog
6,1/1/2012 6:00,-1.5,-3.1,89,7,6.4,101.29,Fog
...,...,...,...,...,...,...,...,...
8716,12/29/2012 4:00,-16.0,-17.2,90,6,9.7,101.25,Fog
8717,12/29/2012 5:00,-14.8,-15.9,91,4,6.4,101.25,Fog
8718,12/29/2012 6:00,-13.8,-15.3,88,4,9.7,101.25,Fog
8719,12/29/2012 7:00,-14.8,-16.4,88,7,8.0,101.22,Fog


# Q. 14) Find all instances when 'Weather is Clear' or 'Visibility is above 40'.

In [ ]:
df[(df.weather_conditions == 'Clear') | (df.visibility_km > 40)]

,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa,weather_conditions
67,1/3/2012 19:00,-16.9,-24.8,50,24,25.0,101.74,Clear
106,1/5/2012 10:00,-6.0,-10.0,73,17,48.3,100.45,Mainly Clear
107,1/5/2012 11:00,-5.6,-10.2,70,22,48.3,100.41,Mainly Clear
108,1/5/2012 12:00,-4.7,-9.6,69,20,48.3,100.38,Mainly Clear
109,1/5/2012 13:00,-4.4,-9.7,66,26,48.3,100.40,Mainly Clear
...,...,...,...,...,...,...,...,...
8749,12/30/2012 13:00,-12.4,-16.2,73,37,48.3,100.92,Mostly Cloudy
8750,12/30/2012 14:00,-11.8,-16.1,70,37,48.3,100.96,Mainly Clear
8751,12/30/2012 15:00,-11.3,-15.6,70,32,48.3,101.05,Mainly Clear
8752,12/30/2012 16:00,-11.4,-15.5,72,26,48.3,101.15,Mainly Clear


# Q. 15) Find all instances when :
### A. 'Weather is Clear' and 'Relative Humidity is greater than 50'
### or
### B. 'Visibility is above 40'

In [181]:
df[(df.weather_conditions == 'Clear') & (df['rel_hum_%'] > 50) | (df['visibility_km'] > 40) ]

,date/time,temp_c,dew_point_temp_c,rel_hum_%,wind_speed_km/h,visibility_km,press_kpa,weather_conditions
106,1/5/2012 10:00,-6.0,-10.0,73,17,48.3,100.45,Mainly Clear
107,1/5/2012 11:00,-5.6,-10.2,70,22,48.3,100.41,Mainly Clear
108,1/5/2012 12:00,-4.7,-9.6,69,20,48.3,100.38,Mainly Clear
109,1/5/2012 13:00,-4.4,-9.7,66,26,48.3,100.40,Mainly Clear
110,1/5/2012 14:00,-5.1,-10.7,65,22,48.3,100.46,Mainly Clear
...,...,...,...,...,...,...,...,...
8749,12/30/2012 13:00,-12.4,-16.2,73,37,48.3,100.92,Mostly Cloudy
8750,12/30/2012 14:00,-11.8,-16.1,70,37,48.3,100.96,Mainly Clear
8751,12/30/2012 15:00,-11.3,-15.6,70,32,48.3,101.05,Mainly Clear
8752,12/30/2012 16:00,-11.4,-15.5,72,26,48.3,101.15,Mainly Clear
